Add support for running DebugConfigurations with serverReadyAction

WardenGnaw opened the pull request:
This PR adds 'config' as an option for `serverReadyAction`. It runs similar to 'name', but if a configuration is generated on the fly, its easier to inject the debug configurations with the serverReadyAction. This modifies the startNamedDebugSession to startDebugSession that takes in a vscode.DebugConfiguration.

Resolves: #186455

roblourens commented:
Did copilot write that 🤔

WardenGnaw commented:
The second half, yes 😅

I wanted to see what it would generate.

roblourens commented on extensions/debug-server-ready/src/extension.ts:
```diff
@@ -246,6 +251,38 @@ class ServerReadyDetector extends vscode.Disposable {
 		});
 	}
 
+	private async startConfigDebugSession(session: vscode.DebugSession, config: vscode.DebugConfiguration) {
```
Any way to share logic with `startNamedDebugSession`?

WardenGnaw replied:
I had one throught:

```
	private async startConfigDebugSession(session: vscode.DebugSession, name: string, config?: vscode.DebugConfiguration) {
		const args = session.configuration.serverReadyAction as ServerReadyAction;
		if (!args.killOnServerStop) {
			await vscode.debug.startDebugging(session.workspaceFolder, config ?? name);
			return;
		}

		const cts = new vscode.CancellationTokenSource();
		const newSessionPromise = this.catchStartedDebugSession(x => x.name === name, cts.token);

		if (!await vscode.debug.startDebugging(session.workspaceFolder, config ?? name)) {
			cts.cancel();
			cts.dispose();
			return;
		}

		const createdSession = await newSessionPromise;
		cts.dispose();

		if (!createdSession) {
			return;
		}

		const stopListener = this.onDidSessionStop(async () => {
			stopListener.dispose();
			this.disposables.delete(stopListener);
			await vscode.debug.stopDebugging(createdSession);
		});
		this.disposables.add(stopListener);
	}
```

Thoughts?

roblourens replied:
Fine with me

WardenGnaw commented:
The new AI generated summary with the new commit

-------------------
This pull request to `extensions/debug-server-ready` adds a new feature that allows a debug configuration to be launched when the server is ready. The changes include modifications to the `startNamedDebugSession` method in `extension.ts` to use a new `startDebugSession` method that can take a debug configuration object as an argument, in addition to the debug configuration name. A new `config` property has been added to the `ServerReadyAction` interface to represent a debug configuration object that can be launched when the server is ready. These changes improve the flexibility and customization of debug sessions.

* `extensions/debug-server-ready/src/extension.ts`: Modified the `startNamedDebugSession` method to use a new `startDebugSession` method that can take a debug configuration object as an argument, in addition to the debug configuration name. A new `config` property has been added to the `ServerReadyAction` interface to represent a debug configuration object that can be launched when the server is ready. (Fb352528)
* <a href="diffhunk://#diff-862a879e83f24473ac5c7196cf1671d006f3e3ab8f60837f079fb22f5cd9e442R161-R204">`extensions/debug-server-ready/package.json`</a>: Added a new debug configuration that can be launched when the server is ready, with properties such as configuration type, request type, and session termination included in the `serverReadyAction` array.

WardenGnaw commented:
@roblourens Is there anything else thats needs to be changed to get this PR merged?

rzhao271 commented on extensions/debug-server-ready/package.json:
```diff
@@ -158,6 +158,50 @@
+                      "action": {
+                        "type": "string",
+                        "enum": [
+                          "startDebugging"
```
Does this mean the only valid option for now is startDebugging?

roblourens replied:
This is an extension of `serverReadyAction` to support a different type when using `startDebugging`. There are other options (the json schema is hard to parse here but they are there). `startDebugging` is the only one that is relevant in this scenario.
